File: 23. Merge k Sorted Lists/23. Merge k Sorted Lists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode[]} lists
 * @return {ListNode}
 */
var mergeKLists = function(lists) {
    if(lists.length==1){
        return lists[0]
    }else if(lists.length<1){
        return null
}
    let middle = Math.floor(lists.length/2)
    let arraySize = lists.length
    let leftSorted = mergeKLists(lists.slice(0,middle))
    let rightSorted = mergeKLists(lists.slice(middle,arraySize))
    let merged = merge(leftSorted,rightSorted)

    return merged

};

let merge = (listOne,listTwo)=>{

    let ptrOne= listOne
    let ptrTwo = listTwo
    let dummyNode = new ListNode()
    let tail = dummyNode;


    while(ptrOne !==null && ptrTwo !==null){
        if(ptrOne.val<ptrTwo.val){
            tail.next=ptrOne
            ptrOne=ptrOne.next
        }else {
            tail.next = ptrTwo
            ptrTwo=ptrTwo.next
        }
        tail=tail.next
    }
    
    while(ptrOne!==null){
        tail.next=ptrOne
        tail=tail.next
        ptrOne= ptrOne.next
    }

    while(ptrTwo!==null){
        tail.next = ptrTwo
        tail= tail.next
        ptrTwo=ptrTwo.next
    }

    return dummyNode.next

}

File: 23. Merge k Sorted Lists/23. Merge k Sorted Lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        
        if len(lists) == 1:
            return lists[0]
        elif len(lists) < 1:
            return None

        left_sorted = self.mergeKLists(lists[:len(lists) // 2])
        right_sorted = self.mergeKLists(lists[len(lists) // 2: len(lists)])
        merged = self.merge(left_sorted, right_sorted)

        return merged

    def merge(self,list_one, list_two):

        ptr_one = list_one
        ptr_two = list_two
        dummy_node = ListNode()
        tail = dummy_node

        while ptr_one is not None and ptr_two is not None:
            if ptr_one.val < ptr_two.val:
                tail.next = ptr_one
                ptr_one = ptr_one.next
            else:
                tail.next = ptr_two
                ptr_two = ptr_two.next
            tail = tail.next

        if(ptr_one):
            tail.next=ptr_one
        elif(ptr_two):
            tail.next=ptr_two

        return dummy_node.next
